src/loaders/anndata-loaders/BaseAnnDataLoader.js

```diff
@@ -0,0 +1,133 @@
+ * like loading cell names and ids. It inherits from AbstractLoader.
+ */
+export default class BaseAnnDataLoader extends AbstractLoader {
+  constructor(params) {
+    super(params);
+
+    // TODO: Use this.requestInit to provide headers, tokens, etc.
+    // eslint-disable-next-line no-unused-vars
+    const { url, requestInit } = this;
+    this.store = new HTTPStore(url);
+  }
+
+  /**
+   * Class method for decoding text arrays from zarr.
+   * @returns {string[]} An array of strings.
+   */
+  // eslint-disable-next-line class-methods-use-this
+  decodeTextArray(buffer) {
+    return (
+      new TextDecoder()
+        .decode(buffer)
+        // https://stackoverflow.com/questions/11159118/incorrect-string-value-xef-xbf-xbd-for-column
+        // for information on the right hand side of the | in the regex.
+        // eslint-disable-next-line no-control-regex
+        .replace(/[\u0000-\u001c]|�c/g, ',')
+        .split(',')
+        .filter(Boolean)
+    );
+  }
+
+  /**
+   * Class method for loading cell set ids.
+   * Takes the location as an argument because this is shared across objects,
+   * which have different ways of specifying location.
+   * @param {string} cellSetZarrLocation A string like obs.leiden or obs.bulk_labels.
+   * @returns {Promise} A promise for an array of ids with one per cell.
+   */
+  loadCellSetIds(cellSetZarrLocation) {
+    const { url } = this;
+    if (this.cellSets) {
+      return this.cellSets;
+    }
+    this.cellSets = Promise.all(
+      cellSetZarrLocation.map(async (setName) => {
+        const res = await fetch(
+          `${this.url}/${setName.replace('.', '/')}/.zattrs`,
+        );
+        const { categories } = await res.json();
+        const categoriesValuesArr = await openArray({
+          store: `${url}/obs/${categories}`,
+          mode: 'r',
+        });
+        const categoriesBuffer = await categoriesValuesArr.compressor.decode(
+          new Uint8Array(await categoriesValuesArr.store.getItem('0')),
+        );
+        const categoriesValues = this.decodeTextArray(categoriesBuffer);
+        const cellSetsArr = await openArray({
+          store: `${url}/${setName.replace('.', '/')}`,
+          mode: 'r',
+        });
+        const cellSetsValues = await cellSetsArr.get();
+        const { data } = cellSetsValues;
+        const mappedCellSetValues = new Array(...data).map(
+          i => categoriesValues[i],
+        );
+        return mappedCellSetValues;
+      }),
+    );
+    return this.cellSets;
+  }
+
+  /**
+   * Class method for loading general numeric arrays.
+   * @param {string} path A string like obsm.X_pca.
+   * @returns {Promise} A promise for a zarr array containing the data.
+   */
+  loadNumeric(path) {
+    const { store } = this;
+    return openArray({
+      store,
+      path: path.replace('.', '/'),
+      mode: 'r',
+    }).then(
+      arr => new Promise((resolve) => {
+        arr.get().then(resolve);
+      }),
+    );
+  }
```

Comment:
`ZarrArray.get` returns a promise, so this can probably be:

```javascript
  loadNumeric(path) {
    const { store } = this;
    return openArray({
      store,
      path: path.replace('.', '/'),
      mode: 'r',
    }).then(arr => arr.get());
  }
```